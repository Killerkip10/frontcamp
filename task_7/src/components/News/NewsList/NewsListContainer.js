import React, { useCallback, useEffect } from 'react';
import { connect } from 'react-redux';

import { NewsList } from './NewsList';
import { getNewsList } from './NewsListEpicActions';
import { changeTopic } from './NewsListActions';

export const NewsListContainerComponent = ({ history, isFetching, news, topic, getNews, changeTopic }) => {
  useEffect(() => { getNews(topic); }, [topic]);

  const handleDetailsClick = useCallback((id, topic) => history.push(`/news/${id}/${topic}`), [history]);
  
  if (isFetching) {
    return <div>...Loading</div>;
  }

  return (
    <NewsList
      news={news}
      topic={topic}
      changeTopic={changeTopic}
      clickDetails={handleDetailsClick}
    />
  );
};

const mapStateToProps = state => ({
  isFetching: state.news.newsList.isFetching,
  news: state.news.newsList.news,
  topic: state.news.newsList.topic,
});

const mapStateToDispatch = dispatch => ({
  getNews: topic => dispatch(getNewsList(topic)),
  changeTopic: topic => dispatch(changeTopic(topic)),
});

const connector = connect(mapStateToProps, mapStateToDispatch);

export const NewsListContainer = connector(NewsListContainerComponent);