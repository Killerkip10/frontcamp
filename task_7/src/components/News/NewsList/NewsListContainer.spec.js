import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow, mount } from 'enzyme';

import { NewsListContainerComponent } from './NewsListContainer';

describe('NewsListContainerComponent', () => {
    let props;
    let useEffect;

    const mockedUseEffect = () => useEffect.mockImplementationOnce(f => f());

    beforeEach(() => {
        props = {
            history: {},
            isFetching: false,
            news: [{
                title: 'title-1',
                url: 'url-1',
            }],
            topic: 'topic-1',
            getNews: jest.fn(),
            changeTopic: jest.fn(),
        };

        useEffect = jest.spyOn(React, 'useEffect');
    });

    it('should render NewsListContainerComponent', () => {
        const component = shallow(<NewsListContainerComponent {...props} />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('should render loader', () => {
        const component = shallow(<NewsListContainerComponent {...props} isFetching />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('should call getNews', () => {
        mount(<NewsListContainerComponent {...props} />);

        expect(props.getNews).toBeCalledWith(props.topic);
    });
});