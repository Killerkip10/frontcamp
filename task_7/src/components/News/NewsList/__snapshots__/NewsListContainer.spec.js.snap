// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsListContainerComponent should render NewsListContainerComponent 1`] = `
<Memo()
  changeTopic={[MockFunction]}
  clickDetails={[Function]}
  news={
    Array [
      Object {
        "title": "title-1",
        "url": "url-1",
      },
    ]
  }
  topic="topic-1"
/>
`;

exports[`NewsListContainerComponent should render loader 1`] = `
<div
  data-cy="loading"
>
  ...Loading
</div>
`;
